### PR TITLE
Simplify inline assembly resolver.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
(hopefully) fixed.

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -666,17 +666,18 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
...
+		bool requiresStorage = identifierInfo.isSlot || identifierInfo.isOffset;
 		if (auto var = dynamic_cast<VariableDeclaration const*>(declaration))
 		{
 			solAssert(var->type(), "Expected variable type!");
 			if (var->immutable())
 			{
 				m_errorReporter.typeError(3773_error, _identifier.location, "Assembly access to immutable variables is not supported.");
-				return numeric_limits<size_t>::max();
+				return false;
 			}
```
Comment (a3d4) [inline libsolidity/analysis/TypeChecker.cpp:681]: `auto watcher = m_errorReporter.errorWatcher();`?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:681]: I would need to change some `if`s into `else if`s but it seems doable!

**Comment by chriseth** [comment]
Updated. It got shorter but also more difficult to review. Your choice! ;)

**Comment by a3d4** [comment]
> Updated. It got shorter but also more difficult to review. Your choice! ;)

Hm... Here `errorWatcher` makes things different, but not necessarily better. Would c9a75c4 from PR #9206 be simpler?  It is also similar to several other cases in `TypeChecker` where we use lambdas to get error descriptions.

I was not sure if I should push to your PR, so made my own. Feel free to close it, if you want.

**Comment by bshastry** [comment]
There are two PRs for this now: this and #9206 . I'm just wondering which one to review :-)

**Comment by a3d4** [comment]
@bshastry

> There are two PRs for this now: this and #9206 . I'm just wondering which one to review :-)

This one :).

**Comment by a3d4** [comment]
@chriseth 
> Updated. It got shorter but also more difficult to review. Your choice! ;)

I cannot convince myself that the variant with the error watcher is better in the given situation. It indeed shortens code, but lack of early returns makes it harder to maintain.

Also, we have cases where we return `false` without emitting an error. So the first commit (5fdbb5c) is an improvement and might be already sufficient. (Yet we might want to refactor `TypeChecker::visit(InlineAssembly)` more in future, I guess.)

**Comment by chriseth** [comment]
Reset to the first commit only.

**Review by a3d4 — APPROVED** [review]

**Pull request merged**